### PR TITLE
Fixes #103: Convert functional components to pure

### LINKED ISSUE #103: Convert components to React.PureComponent
It is currently a (tiny bit) of mental gymnasics since it is not always clear when to use functional or PureComponent. (Well, it is - the latter has state and lifecycle methods). However, moving between files has some context-switching overhead.

It probably makes sense to convert all functional components to PureComponent - slightly less performant (?), however from a "this is how it is done" perspective it makes understanding and contributing easier.

(Not super important, however something to look at)

### PR DESCRIPTION
**Pull request by yjkimjunior**
#103 Convert components to React.PureComponent

### CODE REVIEW

**Comment by yjkimjunior** [comment]
@jacogr was this what you had in mind?

File: packages/app-rpc/src/index.tsx [hunk 1]
```diff
@@ -17,24 +17,28 @@ import Selection from './Selection';
...
-          <Results
-            key='results'
-            queue={queue}
-          />
-        ]}
-      </QueueConsumer>
-    </div>
-  );
+export default class RpcApp extends React.PureComponent<Props> {
+
```
Comment (jacogr) [inline packages/app-rpc/src/index.tsx]: Unneeded blank line here.

File: packages/ui-app/src/Button/Divider.tsx [hunk 2]
```diff
@@ -8,13 +8,15 @@ import React from 'react';
...
 import classes from '../util/classes';
 
-function ButtonDivider ({ className, style }: DividerProps) {
-  return (
-    <div
-      className={classes('ui button compact mini basic', className)}
-      style={style}
-    />
-  );
+// ({ className, style }: DividerProps) {
```
Comment (jacogr) [inline packages/ui-app/src/Button/Divider.tsx]: Superflous comment.

File: packages/ui-app/src/Button/Divider.tsx [hunk 3]
```diff
@@ -8,13 +8,15 @@ import React from 'react';
...
+export default class ButtonDivider extends React.Component<DividerProps> {
+  render () {
+    const { className, style } = this.props;
+    return (
```
Comment (jacogr) [inline packages/ui-app/src/Button/Divider.tsx]: Probably could use a blank line between var declarations & return, doesn't cost anything and aids in readability.

File: packages/app-rpc/src/index.tsx (hunk 1 above)
Comment (jacogr) [inline packages/app-rpc/src/index.tsx]: Unneeded blank line.

File: packages/app-explorer/src/BlockHeader/index.tsx [hunk 4]
```diff
@@ -20,52 +20,54 @@ type Props = I18nProps & {
   value?: Header
 };
 
-function BlockHeader ({ className, value, style }: Props) {
-  if (!value) {
-    return null;
-  }
+class BlockHeader extends React.PureComponent<Props> {
+  render () {
+    const { className, value, style } = this.props;if (!value) {
```
Comment (jacogr) [inline packages/app-explorer/src/BlockHeader/index.tsx]: If on next line. (Pad var blcks with space afterwards)

File: packages/app-explorer/src/index.tsx [hunk 5]
```diff
@@ -17,22 +17,25 @@ import translate from './translate';
...
-          defaultValue: 'best #'
-        })}
-      />
-      <BestHash />
-      <BlockHeaders />
-    </div>
-  );
+class ExplorerApp extends React.PureComponent<Props> {
+  render () {
+    const { className, style, t } = this.props;
```
Comment (jacogr) [inline packages/app-explorer/src/index.tsx]: Blank line after var declarations.

File: packages/app-extrinsics/src/Nonce.tsx [hunk 6]
```diff
@@ -17,22 +17,25 @@ type Props = BareProps & {
...
-    <div
-      className={classes('ui--row', className)}
-      style={style}
-    >
-      <Labelled
-        className='small'
-        label={label}
+export default class Nonce extends React.PureComponent<Props> {
+  render () {
+    const { className, label, style, onChange, value } = this.props;
```
Comment (jacogr) [inline packages/app-extrinsics/src/Nonce.tsx]: Blank line.

File: packages/app-extrinsics/src/index.tsx [hunk 7]
```diff
@@ -16,17 +16,20 @@ import Selection from './Selection';
...
-      <QueueConsumer>
-        {({ queueAdd }: QueueProps) => (
-          <Selection queueAdd={queueAdd} />
-        )}
-      </QueueConsumer>
-    </div>
-  );
+export default class ExtrinsicsApp extends React.PureComponent<Props> {
+  render () {
+    const { className, style } = this.props;
```
Comment (jacogr) [inline packages/app-extrinsics/src/index.tsx]: Followed by blank line.

File: packages/app-rpc/src/Results.tsx [hunk 8]
```diff
@@ -17,34 +17,37 @@ type Props = BareProps & {
...
-    .filter(({ error, result }) =>
-      !isUndefined(error) || !isUndefined(result)
-    )
-    .reverse();
-
-  if (!filtered.length) {
-    return null;
+export default class Results extends React.PureComponent<Props> {
+  render () {
+    const { className, queue = [], style } = this.props;
```
Comment (jacogr) [inline packages/app-rpc/src/Results.tsx]: Followed by blank line.

File: packages/app-rpc/src/index.tsx [hunk 9]
```diff
@@ -17,24 +17,28 @@ import Selection from './Selection';
...
+  render () {
+    const { className, style } = this.props;
+    return (
```
Comment (jacogr) [inline packages/app-rpc/src/index.tsx]: Blank line following var declarations - just helps make difference between areas clearer.

File: packages/app-storage/src/Queries.tsx [hunk 10]
```diff
@@ -17,25 +17,28 @@ type Props = I18nProps & {
   value?: Array<StorageQuery>
 };
 
-function Queries ({ className, onRemove, style, value }: Props) {
-  if (!value || !value.length) {
-    return null;
+class Queries extends React.PureComponent<Props> {
+  render () {
+    const { className, onRemove, style, value } = this.props;
```
Comment (jacogr) [inline packages/app-storage/src/Queries.tsx]: Blank line.

File: packages/apps/src/NodeInfo.tsx [hunk 11]
```diff
@@ -20,24 +20,27 @@ function updateTestInfo (chain?: string) {
...
 
-function NodeInfo ({ className, style, t }: Props) {
-  return (
-    <div className={classes('apps--NodeInfo', className)}>
-      <Chain
-        label={t('info.chain', {
-          defaultValue: 'chain: '
+class NodeInfo extends React.PureComponent<Props> {
+  render () {
+    const { className, style, t } = this.props;
```
Comment (jacogr) [inline packages/apps/src/NodeInfo.tsx]: Blank line.

File: packages/apps/src/NotFound/index.tsx [hunk 12]
```diff
@@ -14,17 +14,20 @@ import translate from '../translate';
...
-      style={style}
-    >
-      {t('notfound.error', {
-        defaultValue: 'ERROR: You have tried to access an application that does not exist'
-      })}
-    </div>
-  );
+class NotFound extends React.PureComponent<Props> {
+  render () {
+    const { className, style, t } = this.props;
```
Comment (jacogr) [inline packages/apps/src/NotFound/index.tsx]: Blank.

File: packages/apps/src/SideBar/Item.tsx [hunk 13]
```diff
@@ -17,17 +17,20 @@ type Props = I18nProps & {
...
-        exact={isExact}
-        to={path || `/${name}`}
-      >
-        <Icon name={icon} /> {t(`sidebar.${name}`, i18n)}
-      </NavLink>
-    </Menu.Item>
-  );
+export default class Item extends React.PureComponent<Props> {
+  render () {
+    const { route: { i18n, icon, isExact, name, path = '' }, t } = this.props;
```
Comment (jacogr) [inline packages/apps/src/SideBar/Item.tsx]: Blank line.

File: packages/apps/src/SideBar/index.tsx [hunk 14]
```diff
@@ -20,51 +20,54 @@ type Props = I18nProps & {
...
-    <div
-      className={classes('apps--SideBar', className)}
-      style={style}
-    >
-      <Menu
-        secondary
-        vertical
+class SideBar extends React.PureComponent<Props> {
+  render () {
+    const { children, className, style, t } = this.props;
```
Comment (jacogr) [inline packages/apps/src/SideBar/index.tsx]: Blank line.

File: packages/ui-app/src/Button/Or.tsx [hunk 15]
```diff
@@ -15,20 +15,24 @@ type Props = I18nProps & {
...
-            defaultValue: 'or'
-          })
-          : text
-      }
-    />
-  );
+// ({ className, style, t, text }: Props)
+class ButtonOr extends React.PureComponent<Props> {
+  render () {
+    const { className, style, t, text } = this.props;
```
Comment (jacogr) [inline packages/ui-app/src/Button/Or.tsx]: Blank line after const, before return.

File: packages/ui-app/src/Button/index.tsx [hunk 16]
```diff
@@ -14,32 +14,37 @@ import isUndefined from '@polkadot/util/is/undefined';
...
-  return isUndefined(text) && isUndefined(children)
-    ? (
-      <SUIButton {...props} />
-    )
-    : (
-      <SUIButton {...props}>
-        {text}{children}
-      </SUIButton>
-    );
+// ({ children, className, floated, icon, isBasic = false, isCircular = false, isDisabled = false, isNegative = false, isPrimary = false, onClick, size, style, text }: ButtonProps) {
```
Comment (jacogr) [inline packages/ui-app/src/Button/index.tsx]: Could be removed.

File: packages/ui-app/src/CopyButton.tsx [hunk 17]
```diff
@@ -18,17 +18,20 @@ type Props = BareProps & {
...
-        isCircular={isCircular}
-        isPrimary={isPrimary}
-        size={size}
-        style={style}
-      />
-    </CopyToClipboard>
-  );
+export default class CopyButton extends React.PureComponent<Props> {
+  render () {
+    const { className, icon = 'copy', isCircular = true, isPrimary = true, size = 'tiny', style, value } = this.props;
```
Comment (jacogr) [inline packages/ui-app/src/CopyButton.tsx]: Extra spacing around var block needed.

File: packages/ui-app/src/InputExtrinsic/SelectSection.tsx [hunk 18]
```diff
@@ -23,22 +23,26 @@ type Props = I18nProps & {
...
-      options={options}
-      style={style}
-      value={section}
-      withLabel={withLabel}
-    />
-  );
+class SelectSection extends React.PureComponent<Props> {
+
+  render () {
+    const { className, defaultValue, isError, label = '', onChange, options, style, t, value: { section }, withLabel } = this.props;
```
Comment (jacogr) [inline packages/ui-app/src/InputExtrinsic/SelectSection.tsx]: Extra blank line needed.

File: packages/ui-app/src/InputExtrinsic/SelectSection.tsx [hunk 19]
```diff
@@ -23,22 +23,26 @@ type Props = I18nProps & {
...
-      })}
-      onChange={onChange}
-      options={options}
-      style={style}
-      value={section}
-      withLabel={withLabel}
-    />
-  );
+class SelectSection extends React.PureComponent<Props> {
+
```
Comment (jacogr) [inline packages/ui-app/src/InputExtrinsic/SelectSection.tsx]: Unneeded blank line here before render

File: packages/ui-app/src/InputRpc/SelectMethod.tsx [hunk 20]
```diff
@@ -24,29 +24,32 @@ type Props = I18nProps & {
   withLabel?: boolean
 };
 
-function SelectMethod ({ className, isError, label = '', onChange, options, style, t, value: { name, section }, withLabel }: Props) {
-  if (!options.length) {
-    return null;
+class SelectMethod extends React.PureComponent<Props> {
+  render () {
+    const { className, isError, label = '', onChange, options, style, t, value: { name, section }, withLabel } = this.props;
```
Comment (jacogr) [inline packages/ui-app/src/InputRpc/SelectMethod.tsx]: Extra blank line needed.

File: packages/ui-app/src/InputRpc/SelectSection.tsx [hunk 21]
```diff
@@ -23,22 +23,25 @@ type Props = I18nProps & {
...
-      onChange={onChange}
-      options={options}
-      style={style}
-      value={section}
-      withLabel={withLabel}
-    />
-  );
+class SelectSection extends React.PureComponent<Props> {
+  render () {
+    const { className, defaultValue, isError, label, onChange, options, style, t, value: { section }, withLabel } = this.props;
```
Comment (jacogr) [inline packages/ui-app/src/InputRpc/SelectSection.tsx]: Extra blank line needed.

File: packages/ui-app/src/InputStorage/SelectKey.tsx [hunk 22]
```diff
@@ -24,29 +24,32 @@ type Props = I18nProps & {
   withLabel?: boolean
 };
 
-function SelectKey ({ className, isError, label = '', onChange, options, style, t, value: { name, section }, withLabel }: Props) {
-  if (!options.length) {
-    return null;
+class SelectKey extends React.PureComponent<Props> {
+  render () {
+    const { className, isError, label = '', onChange, options, style, t, value: { name, section }, withLabel } = this.props;
```
Comment (jacogr) [inline packages/ui-app/src/InputStorage/SelectKey.tsx]: Break between blocks, vars and logic.

File: packages/ui-app/src/InputStorage/SelectSection.tsx [hunk 23]
```diff
@@ -23,22 +23,25 @@ type Props = I18nProps & {
...
-      onChange={onChange}
-      options={options}
-      style={style}
-      value={section}
-      withLabel={withLabel}
-    />
-  );
+class SelectSection extends React.PureComponent<Props> {
+  render () {
+    const { className, defaultValue, isError, label, onChange, options, style, t, value: { section }, withLabel } = this.props;
```
Comment (jacogr) [inline packages/ui-app/src/InputStorage/SelectSection.tsx]: Additional blank line after.

File: packages/ui-app/src/Labelled.tsx [hunk 24]
```diff
@@ -20,22 +20,26 @@ const defaultLabel: any = (// node?
   <div>&nbsp;</div>
 );
 
-export default function Labelled ({ className, children, isHidden = false, label = defaultLabel, style, withLabel = true }: Props) {
-  if (isHidden) {
-    return null;
-  }
+export default class Labelled extends React.PureComponent<Props> {
+
```
Comment (jacogr) [inline packages/ui-app/src/Labelled.tsx]: Unneeded blank line spacing.

File: packages/ui-app/src/Labelled.tsx [hunk 25]
```diff
@@ -20,22 +20,26 @@ const defaultLabel: any = (// node?
...
+  render () {
+    const { className, children, isHidden = false, label = defaultLabel, style, withLabel = true } = this.props;
```
Comment (jacogr) [inline packages/ui-app/src/Labelled.tsx]: Needs some blank line spacing :)

File: packages/ui-app/src/Output.tsx [hunk 26]
```diff
@@ -20,26 +20,29 @@ type Props = BareProps & {
...
-              <CopyButton className='ui--output-button' value={value} />
-            )
-            : null
-        }
-      </div>
-    </Labelled>
-  );
+export default class Output extends React.PureComponent<Props> {
+  render () {
+    const { className, children, isError = false, isHidden, label, style, value, withCopy = false, withLabel } = this.props;
```
Comment (jacogr) [inline packages/ui-app/src/Output.tsx]: Line break after.

File: packages/ui-app/src/Static.tsx [hunk 27]
```diff
@@ -16,19 +16,23 @@ type Props = BareProps & {
...
-      withLabel={withLabel}
-    >
-      <div className='ui dropdown selection disabled'>
-        {value}
-        {children}
-      </div>
-    </Labelled>
-  );
+export default class Static extends React.PureComponent<Props> {
+
```
Comment (jacogr) [inline packages/ui-app/src/Static.tsx]: Unneeded line break.

File: packages/ui-app/src/Static.tsx [hunk 28]
```diff
@@ -16,19 +16,23 @@ type Props = BareProps & {
...
+  render () {
+    const { className, children, isHidden, label, style, value, withLabel } = this.props;
```
Comment (jacogr) [inline packages/ui-app/src/Static.tsx]: Needs a line break.

File: packages/ui-react/src/Container/Demo.tsx [hunk 29]
```diff
@@ -13,26 +13,30 @@ type DemoProps = {
...
-        {`<${component}\n  ${attrsText}\n/>`}
-      </pre>
-      {children}
-    </Container>
-  );
+        return `${opt ? '[' : ''}${attr}${bool ? '' : '='}${bool ? '' : value}${opt ? ']' : ''}`;
+      })
+      .sort()
+      .join('\n  ');
+    return (
```
Comment (jacogr) [inline packages/ui-react/src/Container/Demo.tsx]: Break before. (The last const declaration is dense - nothing changed from original, but not very maintainable)

File: packages/ui-react/src/Container/Demo.tsx [hunk 30]
```diff
@@ -13,26 +13,30 @@ type DemoProps = {
...
 
-export default function Demo ({ attrs, attrsReq, children, component }: DemoProps) {
-  const allAttrs = Object.assign({ className: '"..."', style: '{...}' }, attrs);
-  const attrsText = Object
-    .keys(allAttrs)
-    .map((attr) => {
-      const value = allAttrs[attr];
-      const bool = value === 'bool';
-      const opt = bool || !attrsReq.includes(attr);
+// ({ attrs, attrsReq, children, component }: DemoProps) {
```
Comment (jacogr) [inline packages/ui-react/src/Container/Demo.tsx]: Unneeded comment

File: packages/ui-react/src/Container/index.tsx [hunk 31]
```diff
@@ -11,14 +11,16 @@ import React from 'react';
...
-export default function Container ({ children, className, style }: Props) {
-  return (
-    <div
-      className={['ui--Container', className].join(' ')}
-      style={style}
-    >
-      {children}
-    </div>
-  );
+ // ({ children, className, style }: Props) {
```
Comment (jacogr) [inline packages/ui-react/src/Container/index.tsx]: Unneeded comment.

File: packages/ui-react/src/IdentityIcon/index.tsx [hunk 32]
```diff
@@ -15,12 +15,15 @@ type Props = BaseProps & {
...
-  return (
-    <div
-      className={['ui--IdentityIcon', className].join(' ')}
-      ref={appendIcon(value, size)}
-      style={style}
-    />
-  );
+export default class IdentityIcon extends React.PureComponent<Props> {
+  render () {
+    const { className, size = 64, style, value } = this.props;
```
Comment (jacogr) [inline packages/ui-react/src/IdentityIcon/index.tsx]: Additional blank line.

File: packages/ui-react/src/IdentityIcon/index.tsx [hunk 33]
```diff
@@ -15,12 +15,15 @@ type Props = BaseProps & {
...
+    return (
+      <div
+        className={['ui--IdentityIcon', className].join(' ')}
+        ref={appendIcon(value, size)}
```
Comment (jacogr) [inline packages/ui-react/src/IdentityIcon/index.tsx]: I think this way of appending is probably a bit of an anti-pattern and/or wrong - it works, just not 100%. The append should probably be in this class and done in a more "React-y" way. (Out of scope for this PR maybe, but something to keep in mind)

**Comment by yjkimjunior** [comment]
Thank you for the review @jacogr!

**Pull request merged**